### PR TITLE
Monadify assertions and trace them accordingly

### PR DESCRIPTION
**Pull request by bolt12**

### CODE REVIEW

File: ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs [hunk 1]
```diff
@@ -1165,49 +1189,68 @@ withConnectionManager ConnectionManagerArguments {
...
+                       , Nothing
+                       )
 
       traverse_ (traceWith trTracer . TransitionTrace peerAddr) mbTransition
       traceCounters stateVar
 
       traverse_ cancel mbThread
-      return result
 
+      whenJust mbAssertion $ \tr' -> do
```
Comment (coot) [inline ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs:1248]: is `tr` already in scope?

File: ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs [hunk 2]
```diff
@@ -1895,14 +1976,23 @@ withConnectionManager ConnectionManagerArguments {
...
+                return (DemoteToColdLocalNoop Nothing st
+                       , Nothing
+                       )
               TerminatedState _handleError ->
-                return (DemoteToColdLocalNoop Nothing (abstractState $ Known connState))
+                return ( DemoteToColdLocalNoop Nothing st
+                       , Nothing
+                       )
+
+      whenJust mbAssertion $ \tr' -> do
```
Comment (coot) [inline ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs:1992]: The same question here.

File: ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Types.hs [hunk 3]
```diff
@@ -800,36 +801,47 @@ connectionManagerErrorFromException x = do
...
+  | TrForbiddenOperation           !peerAddr                !AbstractState
+  | TrPruneConnections             ![peerAddr]
+  | TrConnectionCleanup            !(ConnectionId peerAddr)
+  | TrConnectionTimeWait           !(ConnectionId peerAddr)
+  | TrConnectionTimeWaitDone       !(ConnectionId peerAddr)
+  | TrConnectionManagerCounters    !ConnectionManagerCounters
+  | TrState                        !(Map peerAddr AbstractState)
   -- ^ traced on SIGUSR1 signal, installed in 'runDataDiffusion'
-  | TrUnexpectedlyMissingConnectionState !(ConnectionId peerAddr)
+  | TrCMUnexpectedlyFalseAssertion !(AssertionLocation peerAddr)
```
Comment (coot) [inline ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Types.hs:844]:
```suggestion
  | TrUnexpectedlyFalseAssertion !(AssertionLocation peerAddr)
```
I plan for the types in `connection-manager`, `inbound-governor` etc. to be always imported qualified.

**Review by coot — COMMENTED** [review]
Just minor comments.

File: ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs (hunk 1 above)
Comment (bolt12) [reply ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs:1248]: I'll double check and change it accordingly!

File: ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Types.hs (hunk 3 above)
Comment (bolt12) [reply ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Types.hs:844]: I see, I left the name CM/IG there so one could check in which component was this traced. Is there any way to reliably know without the names?

Comment (coot) [reply ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Types.hs:844]: yes, the `cardano-node` tracer will include tracer name in its context.

Comment (coot) [reply ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Types.hs:844]: I think it is derived from the type, so we might need to check how `ouroboros-consensus` is dealing with this (many tracers there are just called `Tracer`).

**Review by coot — APPROVED** [review]
LGTM

**Comment by bolt12** [comment]
bors merge

**Pull request merged**